src/tty/windows/conpty.rs

```diff
@@ -0,0 +1,285 @@
+// Copyright 2016 Joe Wilm, The Alacritty Project Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{process_should_exit, Pty, HANDLE};
+
+use std::env;
+use std::i16;
+use std::mem;
+use std::os::windows::io::IntoRawHandle;
+use std::ptr;
+use std::sync::Arc;
+
+use dunce::canonicalize;
+use lazy_static::lazy_static;
+use mio_anonymous_pipes::{EventedAnonRead, EventedAnonWrite};
+use miow;
+use widestring::U16CString;
+use winapi::shared::basetsd::{PSIZE_T, SIZE_T};
+use winapi::shared::minwindef::{BYTE, DWORD};
+use winapi::shared::ntdef::{HANDLE, HRESULT, LPCWSTR, LPWSTR};
+use winapi::shared::winerror::S_OK;
+use winapi::um::libloaderapi::{GetModuleHandleA, GetProcAddress};
+use winapi::um::processthreadsapi::{
+    CreateProcessW, InitializeProcThreadAttributeList, UpdateProcThreadAttribute,
+    PROCESS_INFORMATION, STARTUPINFOW,
+};
+use winapi::um::winbase::{EXTENDED_STARTUPINFO_PRESENT, STARTUPINFOEXW};
+use winapi::um::wincontypes::{COORD, HPCON};
+
+use crate::cli::Options;
+use crate::config::{Config, Shell};
+use crate::display::OnResize;
+use crate::term::SizeInfo;
+
+#[allow(non_snake_case)]
+struct ConptyApi {
+    CreatePseudoConsole:
+        unsafe extern "system" fn(COORD, HANDLE, HANDLE, DWORD, *mut HPCON) -> HRESULT,
+    ResizePseudoConsole: unsafe extern "system" fn(HPCON, COORD) -> HRESULT,
+    ClosePseudoConsole: unsafe extern "system" fn(HPCON) -> HRESULT,
+}
+
+lazy_static! {
+    static ref CONPTY_API: Option<ConptyApi> = unsafe {
+        let hmodule = GetModuleHandleA("kernel32\0".as_ptr() as _);
+        assert!(!hmodule.is_null());
+
+        let cpc = GetProcAddress(hmodule, "CreatePseudoConsole\0".as_ptr() as _);
+        let rpc = GetProcAddress(hmodule, "ResizePseudoConsole\0".as_ptr() as _);
+        let clpc = GetProcAddress(hmodule, "ClosePseudoConsole\0".as_ptr() as _);
+
+        if cpc.is_null() || rpc.is_null() || clpc.is_null() {
+            None
+        } else {
+            Some(ConptyApi {
+                CreatePseudoConsole: mem::transmute(cpc),
+                ResizePseudoConsole: mem::transmute(rpc),
+                ClosePseudoConsole: mem::transmute(clpc),
+            })
+        }
+    };
+}
```

Comment:
`lazy_static` is not required, however I chose to use it because it's something that is effectively global (either the symbols will be present in kernel32 when you first try or it's never going to work).

I can easily move this into the Conpty struct if you'd prefer? Each call to new() will then be looking up these symbols (unlikely to be a problem).

Reply:
Is the construction of the Conpty going to happen more than once? Lazy static would need to be called once at startup, I'd assume that's the same for the Conpty?

If that's the case, I'd prefer it moved to use without `lazy_static`. I feel like it's great if we can work our way around it and getting rid of a dependency is never a bad idea either.